### PR TITLE
902 logging in to aemaacs instances through access tokens

### PR DESCRIPTION
**Pull request by jean-khechfe-wttech**

### CODE REVIEW

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt [hunk 1]
```diff
@@ -32,21 +32,21 @@ class FileTransferManager(private val common: CommonExtension) : FileTransfer {
         common.prop.string("fileTransfer.password")?.let { set(it) }
     }
 
+    var bearerToken: String? = null
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:35]: common.obj.string as for other mutable props;


see https://docs.gradle.org/current/userguide/lazy_configuration.html

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt [hunk 2]
```diff
@@ -32,21 +32,21 @@ class FileTransferManager(private val common: CommonExtension) : FileTransfer {
...
 
-    val credentials: Pair<String, String> get() = when {
-        user.isPresent && password.isPresent -> (user.get() to password.get())
-        else -> throw FileTransferException("File transfer credentials are missing!")
-    }
+    val credentials: Pair<String, String>?
+        get() = if (user.isPresent && password.isPresent) user.get() to password.get() else null
 
-    val credentialsString get() = credentials.run { "$first:$second" }
+    val credentialsString get() = credentials?.run { "$first:$second" } ?: ""
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:44]: the empty string is not a solution;

the caller need to check for nullability or availability of that property
Kotlin intentionally has nullable types to check defined values at compile-time; using empty strings is a form of hacking that mechanism

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt [hunk 3]
```diff
@@ -70,6 +70,7 @@ class FileTransferManager(private val common: CommonExtension) : FileTransfer {
     val http = HttpFileTransfer(common).apply {
         client.basicUser.convention(user)
         client.basicPassword.convention(password)
+        client.bearerToken = bearerToken
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:73]: please same as on lines above

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt [hunk 4]
```diff
@@ -75,20 +77,25 @@ open class HttpClient(private val common: CommonExtension) {
         common.prop.string("httpClient.basicPassword")?.let { set(it) }
     }
 
-    var basicCredentials: Pair<String?, String?>
-        get() = when {
-            basicUser.isPresent && basicPassword.isPresent -> (basicUser.get() to basicPassword.get())
-            else -> throw HttpException("HTTP client basic credentials are missing!")
-        }
+    var basicCredentials: Pair<String, String>?
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:80]: but here you have nullable, not empty string :D be consistent :)

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt [hunk 5]
```diff
@@ -75,20 +77,25 @@ open class HttpClient(private val common: CommonExtension) {
...
+            }
         }
 
     fun fileTransferCredentials() {
         basicCredentials = common.fileTransfer.credentials
     }
 
+    var bearerToken: String? = null
+
+    fun fileTransferBearerToken() {
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:95]: this function looks seems to be more `val` than `fun`
maybe also refactor `fileTransferCredentials`

**Review by krystian-panek-wttech — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt (hunk 2 above)
Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:44]: makes sense, fixed, thanks!

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt (hunk 1 above)
Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:35]: didn't occur to me to leave "string { }" empty, Fixed, Thanks!

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt (hunk 4 above)
Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:80]: got it :)

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt (hunk 5 above)
Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:95]: done

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt [hunk 6]
```diff
@@ -32,21 +32,21 @@ class FileTransferManager(private val common: CommonExtension) : FileTransfer {
...
     val domain = common.obj.string {
         common.prop.string("fileTransfer.domain")?.let { set(it) }
     }
 
-    val credentials: Pair<String, String> get() = when {
-        user.isPresent && password.isPresent -> (user.get() to password.get())
-        else -> throw FileTransferException("File transfer credentials are missing!")
-    }
+    val credentials: Pair<String, String>?
+        get() = if (user.isPresent && password.isPresent) user.get() to password.get() else null
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:42]: you could make it even more defensive


```
if (!user.orNull.isNullOrBlank() && !password.orNull.isNullOrBlank()) user.get() to password.get() else null
```

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt [hunk 7]
```diff
@@ -127,7 +134,7 @@ open class HttpClient(private val common: CommonExtension) {
         useDefaults()
     }
 
-    fun HttpClientBuilder.useDefaults() {
+    private fun HttpClientBuilder.useDefaults() {
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:137]: this was intentionally public as it is used in some lamba and that lambda could be overidden then reuse `useDefaults()` and add own logic afterwards

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt [hunk 8]
```diff
@@ -140,6 +147,14 @@ open class HttpClient(private val common: CommonExtension) {
             )
         }
 
+        if (!bearerToken.orNull.isNullOrBlank()) {
+            val headers = listOf(
+                BasicHeader(HttpHeaders.CONTENT_TYPE, "application/json"),
```
Comment (krystian-panek-wttech) [inline src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:152]: this is wrong. you cannot add this header here as common plugin is general; not your case-specific

Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:152]: removed

File: src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt (hunk 6 above)
Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/file/transfer/FileTransferManager.kt:42]: done

File: src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt [hunk 9]
```diff
@@ -140,6 +147,13 @@ open class HttpClient(private val common: CommonExtension) {
             )
         }
 
+        if (!bearerToken.orNull.isNullOrBlank()) {
+            val headers = listOf(
+                BasicHeader(HttpHeaders.AUTHORIZATION, "Bearer ${bearerToken.get()}")
+            )
+            setDefaultHeaders(headers)
```
[the remark this answers is not in the document]
Comment (jean-khechfe-wttech) [reply src/main/kotlin/com/cognifide/gradle/common/http/HttpClient.kt:154]: yeah i couldn't find one

**Pull request merged**